runtime/arm64.S

```diff
@@ -150,92 +154,178 @@ G(name):
         .size   G(name), .-G(name)
 #endif
 
+/* Stack switching operations */
+
+/* struct stack_info */
+#define Stack_sp(reg)           [reg]
+#define Stack_exception(reg)    [reg, #8]
+#define Stack_handler(reg)      [reg, #16]
+#define Stack_handler_from_cont(reg) [reg, #15]
+
+/* struct c_stack_link */
+#define Cstack_stack(reg)       [reg]
+#define Cstack_sp(reg)          [reg, #8]
+#define Cstack_prev(reg)        [reg, #16]
+
+/* struct stack_handler */
+#define Handler_value(reg)      [reg]
+#define Handler_exception(reg)  [reg, #8]
+#define Handler_effect(reg)     [reg, #16]
+#define Handler_parent(reg)     [reg, #24]
+
+/* Switch from OCaml to C stack. */
+.macro SWITCH_OCAML_TO_C
+    /* Fill in Caml_state->current_stack->sp */
+        ldr     TMP, Caml_state(current_stack)
+        mov     TMP2, sp
+        str     TMP2, Stack_sp(TMP)
+    /* Fill in Caml_state->c_stack */
+        ldr     TMP2, Caml_state(c_stack)
+        str     TMP, Cstack_stack(TMP2)
+        mov     TMP, sp
+        str     TMP, Cstack_sp(TMP2)
+    /* Switch to C stack */
+        mov     sp, TMP2
+.endm
+
+/* Switch from C to OCaml stack. */
+.macro SWITCH_C_TO_OCAML
+        ldr     TMP, Cstack_sp(sp)
+        mov     sp, TMP
+.endm
+
+/* Save all of the registers that may be in use to a free gc_regs bucket
+   and store ALLOC_PTR and TRAP_PTR back to Caml_state
+   At the end the saved registers are placed in Caml_state(gc_regs)
+ */
+.macro SAVE_ALL_REGS
+    /* First, save the young_ptr & exn_handler */
+        str     ALLOC_PTR, Caml_state(young_ptr)
+        str     TRAP_PTR, Caml_state(exn_handler)
+    /* Now, use TMP to point to the gc_regs bucket */
+        ldr     TMP, Caml_state(gc_regs_buckets)
+        ldr     TMP2, [TMP, 0] /* next ptr */
+        str     TMP2, Caml_state(gc_regs_buckets)
+    /* Save allocatable registers */
+        stp     x0, x1, [TMP, 16]
+        stp     x2, x3, [TMP, 32]
+        stp     x4, x5, [TMP, 48]
+        stp     x6, x7, [TMP, 64]
+        stp     x8, x9, [TMP, 80]
+        stp     x10, x11, [TMP, 96]
+        stp     x12, x13, [TMP, 112]
+        stp     x14, x15, [TMP, 128]
+        stp     x19, x20, [TMP, 144]
+        stp     x21, x22, [TMP, 160]
+        stp     x23, x24, [TMP, 176]
+        str     x25, [TMP, 192]
+        stp     d0, d1, [TMP, 208]
```

Comment:
For the FP registers, we save the caller-save allocatable registers only.  What about a comment?
```suggestion
    /* Save caller-save floating-point registers (callee-saves are preserved by C functions) */
        stp     d0, d1, [TMP, 208]
```